black_it/MCS.R

```diff
@@ -0,0 +1,105 @@
+# Reading the data
+
+mData <- read.csv("Data.csv")
+attach(mData)
+
+fnComplete <- function(x,l=301) {
+	return(c(x,rep(x[length(x)],l-length(x))))
+}
+
+# Building the MCS
+
+vZ <- 300+4.9*(1:301)
+
+mDist <- NULL
+for (i in 1:16) {
+	mDist2 <- matrix(0,nrow=200,ncol=2)
+	mDist2[,2] <- i
+	mDataCOP <- mData[mData$COP==i,]
+	vRun <- unique(mDataCOP$run)
+	dRun0 <- min(vRun)
+	for (j in 1:200) {
+		vX <- fnComplete(mDataCOP[mDataCOP$run==(dRun0-1+j),]$pr.solved2)
+		mDist2[j,1] <- sum(abs(vX-vZ)^2)/(301*1000*1000)
+	}
+	mDist	<- rbind(mDist,mDist2)
+	mDataCOP <- NULL
+}
+mDist <- as.data.frame(mDist)
+names(mDist) <- c("Dist","COP")
+
+library(data.table) 
+mD <- setDT(mDist)[,list(mean=mean(Dist),var=var(Dist)),by=c("COP")]
+
+fnTest <- function(vMean,vVar,iN,dA=1) {
+	iM <- length(vMean)
+	if (iM>2) {
+		mA <- cbind(rep(1,iM-1),diag(-rep(1,iM-1)))
+		mVar <- diag(vVar[-1])
+		mVar2 <- mVar+vVar[1]
+		dTest <- drop(iN*t(mA%*%vMean)%*%solve(mVar2)%*%(mA%*%vMean))
+	} else {
+		dTest <- iN*(vMean[1]-vMean[2])^2/(vVar[1]+vVar[2])
+	}
+	if (iM>1) {
+		dQuan <- qchisq(1-dA,iM-1)
+		dPVa <- 1-pchisq(dTest,iM-1)
+		} else {
+		dQuan <- 0
+		dPVa <- 1
+	}
+	dResult <- 1*(dTest>dQuan)
+	if (is.na(dResult)) dResult <- 0
+	return(list(test=dTest,q=dQuan,p=dPVa,result=dResult))
+}
+
+fnElim	<- function(vMean,vVar,vIndex) {
+	iM <-	which.max(vMean)
+	return(list(mean=vMean[-iM],var=vVar[-iM],indices=vIndex[-iM],elim=vIndex[iM]))
+}
+
+fnMCS <- function(vMean,vVar,vIndex,dA,verbose=1) {
```

Comment:
Same here

Reply:
Hi @AldoGl, I committed a commented version of the MCS code